### PR TITLE
test/runtests.jl: Allow parallel tests for isolated environments

### PR DESCRIPTION
**Pull request by zimoun**
This fixes past issues #43205 or #40348 and implements what comments from PR #40359 committed in fe09ad9e0587a1b2f228aef727f6d1495efea98c had suggested.  It seems an improvement of #29384.

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -91,19 +91,13 @@ prepend!(tests, linalg_tests)
...
-    #   * https://github.com/JuliaLang/julia/pull/29384
-    #   * https://github.com/JuliaLang/julia/pull/40348
     n = 1
-    if net_on
-        n = min(Sys.CPU_THREADS, length(tests))
+    # Allow parallel tests with isolated environment. Some isolated build
+    # environments including Nix or GNU Guix have `net_on` set to false but
+    # still have access to the loopback interface.
+    if net_on || haskey(ENV, "JULIA_CPU_THREADS")
+        x = haskey(ENV, "JULIA_CPU_THREADS") ? parse(Int, ENV["JULIA_CPU_THREADS"]) : Sys.CPU_THREADS
```
Comment (vtjnash) [inline test/runtests.jl:99]: seems odd to go to all this work to compute the value that you already have
```suggestion
        x = Sys.CPU_THREADS
```

Comment (zimoun) [reply test/runtests.jl:99]: Maybe I miss something, but I can use a number of threads different of what my machine physically has.  For instance, one of my machine has 64 cores and I only want to use 32.  This 32 is specified by `JULIA_CPU_THREADS` is I read correctly the doc.

Comment (vtjnash) [reply test/runtests.jl:99]: `JULIA_CPU_THREADS` is already specified as the way to override Sys.CPU_THREADS

Comment (zimoun) [reply test/runtests.jl:99]: Ah indeed.  Thanks.  So only  the `|| haskey(ENV, "JULIA_CPU_THREADS")` seems relevant for building in isolated environment where `net_on` is false but `loopback` is available.

Right?

Comment (Apteryks) [reply test/runtests.jl:99]: It's a hack that came from NixOS (looking at a container-specific environment variable to override the problematic "net_on").  Ideally net_on should be changed to something for the very thing it requires here, which is a loopback interface.

Comment (zimoun) [reply test/runtests.jl:99]: @Apteryks  Yes, but it requires to change `getipaddr` which seems a deep change, where test on `JULIA_CPU_THREADS` seems the documented way to proceed., AFAIU.

```
    net_on = true
    try
        ipa = getipaddr()
    catch
        if ci_option_passed
            @error("Networking unavailable, but `--ci` was passed")
            rethrow()
        end
        net_on = false
        @warn "Networking unavailable: Skipping tests [" * join(net_required_for, ", ") * "]"
        filter!(!in(net_required_for), tests)
    end
```

https://github.com/JuliaLang/julia/blob/master/test/choosetests.jl#L166-L177

Comment (Apteryks) [reply test/runtests.jl:99]: Yes, it's still a good stopgap solution in my opinion.

**Comment by DilumAluthge** [comment]
I don't think that checking for the `JULIA_CPU_THREADS` environment variable is the correct thing to do here. There may be other reasons for that environment variable to be set.

Instead, we should have an environment variable specifically for this purpose.

I'll open a new pull request.

**Comment by zimoun** [comment]
@DilumAluthge I agree that another environment variable should be better than `JULIA_CPU_THREADS`, only for naming IMHO.  However, I miss which are the other reasons and all the cases are covered.  Well, you are interested by the case `net_on=false` and `JULIA_CPU_THREADS != 1` and still run the tests with only one thread?  Why do you tweak `JULIA_CPU_THREADS` thus?

I think this `||` test is correct, at least in agreement with doc of `JULIA_CPU_THREADS`.  Could you detail these other reasons for that environment variable to be set and not effectively set `addprocs_with_testenv` to the number of core specified by user using the environment variable for that purpose `JULIA_CPU_THREADS`?  IIUC.

Otherwise, it is counter intuitive with _Overrides the global variable `Base.Sys.CPU_THREADS`, the number of logical CPU cores available._  At least, counter-intuitive for me. ;-)

**Comment by DilumAluthge** [comment]
Just as an example off the top of my head, a user may have `JULIA_CPU_THREADS` set in their `.bashrc` file for other purposes.

If the idea here is that we want to override the automatic network detection for the purpose of running the tests with multiple worker processes, we should have an environment variable explicitly for this purpose.

**Comment by DilumAluthge** [comment]
#43217

**Comment by zimoun** [comment]
> Just as an example off the top of my head, a user may have `JULIA_CPU_THREADS` set in their `.bashrc` file for other purposes.

Similarly, a user may have `JULIA_TEST_NETWORKING_AVAILABLE` in their `.bashrc` file for other purposes. ;-)


> If the idea here is that we want to override the automatic network detection for the purpose of running the tests with multiple worker processes, we should have an environment variable explicitly for this purpose.

The automatic network detection is not overriden here.  It incorrectly reports the status of `loopback`; as detailed here https://github.com/JuliaLang/julia/issues/43205#issuecomment-977742642.

The correct fix is to fix `getipaddr`.  But it is a deep change.  Where the proposed PR #43211 here is almost trivial and fix the immediate bug waiting a proper fix of `getipaddr`, IMHO.

**Comment by Apteryks** [comment]
Hello,

Dilum Aluthge ***@***.***> writes:

> Just as an example off the top of my head, a user may have
> `JULIA_CPU_THREADS` set in their `.bashrc` file for other purposes.
>
> If the idea here is that we want to override the automatic network
> detection for the purpose of running the tests with multiple worker
> processes, we should have an environment variable explicitly for this
> purpose.

Would it be possible to have another check such has_loopback or
something?  In other words, checking for the minimum needed to run the
tests in parallel with workers.  This would always do the correct thing,
no matter the environment.

Of course, for places where tests try to reach out to the network, the
current check would remain the one to use.

Does it make sense?  Would that be difficult to implement?

Thanks!

**Comment by DilumAluthge** [comment]
> Would it be possible to have another check such has_loopback or something? In other words, checking for the minimum needed to run the tests in parallel with workers.

Sure, I think that would be a good check to have. If someone wants to make a PR adding that check (along with a demonstration that their check works successfully even in environments where `Sockets.getipaddr()` throws an exception), that would be great!

#43217 is just a stop-gap solution until someone makes a PR to add such a check.

**Comment by Apteryks** [comment]
OK, thanks for taking the time to reply!

**Pull request closed without merging**